fix(run-protocol): addVaultType with invalid args modified state

dckc opened the pull request:
closes: #4284

## Description

Check args against a pattern before doing anything that might modify state.

### Security Considerations

fixes a leak

### Testing Considerations

includes a regression test

cc @arirubinstein

Chris-Hibbert commented:
I think the test failures may be due to a github outage issue. I'm seeing similar failures in an unrelated PR.

Chris-Hibbert commented on packages/run-protocol/src/vaultFactory/vaultFactory.js:
```diff
@@ -108,6 +108,24 @@ export const start = async (zcf, privateArgs) => {
+      numerator: amountPattern,
+      denominator: amountPattern,
+    };
+    fit(
```
Why does this help?  (`fit` is new for me.)  As I read the doc for `fit`, it looks like it returns a boolean.  What causes it to act as a guard here?

dckc replied:
The doc for `fit` might be a bit spread out. It's defined as:

```
  const fit = (specimen, patt) => {
    checkMatches(specimen, patt, assertChecker);
  };
```

... where [assertChecker](https://github.com/endojs/endo/blob/master/packages/marshal/src/helpers/passStyle-helpers.js#L50) comes from `@endo/marshal`. The doc for its type, [Checker](https://github.com/endojs/endo/blob/master/packages/marshal/src/types.js#L228-L240) should help:

```
 * @callback Checker
 * Internal to a useful pattern for writing checking logic
 * (a "checkFoo" function) that can be used to implement a predicate
 * (an "isFoo" function) or a validator (an "assertFoo" function).
 *
 *    * A predicate ideally only returns `true` or `false` and rarely throws.
 *    * A validator throws an informative diagnostic when the predicate
 *      would have returned `false`, ...
```

Chris-Hibbert commented on packages/run-protocol/test/vaultFactory/test-vaultFactory.js:
```diff
@@ -2296,3 +2296,52 @@ test('mutable liquidity triggers and interest sensitivity', async t => {
+    // @ts-ignore bad args on purpose for test
+    .addVaultType(aethIssuer, kw, null)
+    .then(oops => t.fail(`${oops}`))
+    .catch(reason1 =>
+      t.throwsAsync(
+        // @ts-ignore bad args on purpose for test
+        E(vaultFactory).addVaultType(aethIssuer, kw, null),
+        { message: reason1.message },
```
This shows that it returns the same error message on the second call, but it doesn't show that it could succeed on the second call if the other args were no longer erroneous.

dckc replied:
Good thing you prompted me to finish the test; it showed that the call to `fix()` didn't work; it was missing a `harden()`.

Chris-Hibbert commented on packages/run-protocol/src/vaultFactory/vaultFactory.js:
```diff
@@ -129,6 +129,24 @@ export const start = async (zcf, privateArgs) => {
 
   /** @type {AddVaultType} */
   const addVaultType = async (collateralIssuer, collateralKeyword, rates) => {
+    const amountPattern = { brand: M.remotable(), value: M.any() };
+    const ratioPattern = {
+      numerator: amountPattern,
+      denominator: amountPattern,
+    };
```
Should these be in some standard place?  Seems like we'll use them often.

dckc replied:
how about `packages/run-protocol/src/contractSupport.js`?

Chris-Hibbert replied:
That seems right.

Chris-Hibbert reviewed (approved):
Looks good.

I recommend finding a common place for soon-to-be common code.

Chris-Hibbert commented on packages/run-protocol/src/contractSupport.js:
```diff
@@ -1,8 +1,15 @@
 
 const { details: X } = assert;
 
+export const amountPattern = { brand: M.remotable(), value: M.any() };
+export const ratioPattern = {
```
They'll be part of a hardened structure, but it would still be a mistake if anyone modified them before using them. Wouldn't it be helpful to `harden()` them?

dckc replied:
eek! yes, of course; lack of `harden()` is a pretty bad hazzard here.

dckc replied:
`harden()` is designed to be used on all newly created objects before they escape the static scope in which they are created:

> `harden()` should be called on all objects that will be transferred across a trust boundary. It's important to `harden()` an object before exposing the object by returning it or passing it to some other function.
 -- https://agoric.com/documentation/guides/js-programming/hardened-js.html#defensive-objects-with-harden

> To enable sound static reasoning, in Jessie all objects made by literal expressions (object literals, array literals, the many forms of function literals) must be tamper-proofed with harden before they can be aliased or escape from their static context of origin.
 -- https://github.com/endojs/Jessie

This supports defensive correctness at a granularity finer than vats -- all the way down to individual objects.

In the interest of programmer convenience, in some of our APIs, the callee hardens for the caller. I think @erights is uneasy in all such cases, since (1) it risks masking bugs / confusion (as discussed in https://agoric.com/documentation/guides/js-programming/far.html#pass-styles-and-harden ) , and (2) some callers might be surprised by having their objects hardened as a consequence of passing them to functions such as `fit()`.

turadg commented on packages/run-protocol/test/vaultFactory/test-vaultFactory.js:
```diff
@@ -34,6 +34,7 @@ import { startVaultFactory } from '../../src/econ-behaviors.js';
 import '../../src/vaultFactory/types.js';
 import * as Collect from '../../src/collect.js';
 import { calculateCurrentDebt } from '../../src/interest-math.js';
+import { M, matches } from '@agoric/store';
```
@dckc Heads up on this lint error.

We really should automate import sorting.

dckc commented:
oops; I squashed locally. I should have let github do that...

dckc commented:
@Chris-Hibbert This use of pattern matching refuses extra fields. For example, `demoIssuers.js` was providing `initialPrice` and `initialMargin` along with `interestRate` and such. That was throwing. I just changed `demoIssuers.js` to not do that, but... is that the way we want to go?

mhofman commented:
> @Chris-Hibbert This use of pattern matching refuses extra fields. For example, `demoIssuers.js` was providing `initialPrice` and `initialMargin` along with `interestRate` and such. That was throwing. I just changed `demoIssuers.js` to not do that, but... is that the way we want to go?

I'm triggering the loadgen CI on this as I believe it will break.

In general I'm against unrecognized props in option bags not being simply ignored.

mhofman reviewed (changes requested):
Requesting change until I figure out the result of the loadgen test

mhofman commented:
> I'm triggering the loadgen CI on this as I believe it will break.
> 
> In general I'm against unrecognized props in option bags not being simply ignored.

Test ran without errors, but I'm still concerned extra properties should cause errors, and that this will bite us in the future

Chris-Hibbert commented:
> In general I'm against unrecognized props in option bags not being simply ignored.

@mhofman 
Sorry too many negatives collided in my head there. Can you say something simple about what you favor?

Chris-Hibbert reviewed (approved):
Thanks for the changes. Once CI and @mhofman are satisfied, this can be merged.

LGTM.

mhofman reviewed (dismissed):
Given this PR is on automerge, I want to make sure the extra property question is resolved before this lands.

dckc commented:
TIL: `M.split()`

@mhofman I added tests for both missing and extra params.
@Chris-Hibbert I hoisted `vaultParamPattern` to `params.js` since it took so much work to get right.

Chris-Hibbert commented:
> TIL: M.split()

Did you find documentation for it somewhere? I perused `.../packages/store` and didn't come up with anything.

Chris-Hibbert commented:
> I hoisted vaultParamPattern to params.js since it took so much work to get right... maybe a client would want to use it?

Seems plausible

> And rather than fit(harden([...args])) I made separate calls to fit() for each arg... except collateralKeyword, where I used 
assertKeywordName(collateralKeyword) instead. And I added a test for bad keyword.

That works.

mhofman commented on packages/run-protocol/test/vaultFactory/test-vaultFactory.js:
```diff
@@ -2190,3 +2191,79 @@ test('mutable liquidity sensitivity of triggers and interest', async t => {
+  const { vaultFactory } = services.vaultFactory;
+
+  const params = { ...defaultParamValues(aethBrand), shoeSize: 10 };
+  const extraParams = { ...params, shoeSize: 10 };
```
Thanks!

mhofman reviewed (approved):
Approving for the handling of extra params

dckc commented:
> > TIL: M.split()
> 
> Did you find documentation for it somewhere? I perused `.../packages/store` and didn't come up with anything.

You too? Now I feel less dumb. :)

With help from @erights , I found the following; he and @mhofman and I agreed that it could be made easier to find:

https://github.com/Agoric/agoric-sdk/blob/4a46d9e37fc3946edb99e8ce41f52872e7816094/packages/store/src/types.js#L530-L536
